File: desktop/src/Plexus.Interop.Broker.Core/Internal/ClientConnectionProcessor.cs
/**
 * Copyright 2017 Plexus Interop Deutsche Bank AG
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
namespace Plexus.Interop.Broker.Internal
{
    using Plexus.Channels;
    using Plexus.Interop.Transport;
    using System;
    using System.Collections.Generic;
    using System.Threading.Tasks;

    internal sealed class ClientConnectionProcessor : IClientConnectionProcessor
    {
        private readonly ILogger _log;

        private readonly IClientConnection _connection;
        private readonly IClientRequestHandler _clientRequestHandler;
        private readonly HashSet<Task> _runnningTasks = new HashSet<Task>();

        public ClientConnectionProcessor(IClientConnection connection, IClientRequestHandler clientRequestHandler)
        {
            _connection = connection;
            Id = _connection.Id;
            _log = LogManager.GetLogger<ClientConnectionProcessor>(Id.ToString());
            _clientRequestHandler = clientRequestHandler;
            Completion = TaskRunner.RunInBackground(ProcessAsync);
        }

        public UniqueId Id { get; }

        public Task Completion { get; }

        private async Task ProcessAsync()
        {
            _log.Debug("Listening for connection requests: {0}", _connection);
            var listenChannelTask = _connection.IncomingChannels.ConsumeAsync((Action<ITransportChannel>)HandleChannel).IgnoreExceptions();
            lock (_runnningTasks)
            {
                _runnningTasks.Add(listenChannelTask);
            }
            await _connection.Completion.ConfigureAwait(false);
            Task completion;
            lock (_runnningTasks)
            {
                completion = Task.WhenAll(_runnningTasks);
            }
            await completion.ConfigureAwait(false);
            _log.Debug("Connection listening completed: {0}", _connection);
        }

        private void HandleChannel(ITransportChannel channel)
        {
            _log.Debug("Processing new channel {0} from connection {1}", channel.Id, _connection);
            var task = TaskRunner.RunInBackground(HandleChannelAsync, channel);
            lock (_runnningTasks)
            {
                _runnningTasks.Add(task);
            }
            task.ContinueWithSynchronously((Action<Task, object>)OnTaskCompleted, channel);
        }

        private async Task HandleChannelAsync(object state)
        {
            ITransportChannel channel = null;
            try
            {
                channel = (ITransportChannel)state;
                await _clientRequestHandler.HandleChannelAsync(_connection, channel).ConfigureAwait(false);
                _log.Debug("Channel {0} completed", channel.Id);
            }
            catch (Exception ex)
            {
                _log.Warn(ex, "Exception on handling channel {0}", channel?.Id);
            }
        }

        private void OnTaskCompleted(Task task, object state)
        {
            lock (_runnningTasks)
            {
                _runnningTasks.Remove(task);
            }
        }
    }
}

File: web/packages/broker/src/broker/ClientConnectionProcessor.ts
/**
 * Copyright 2017-2018 Plexus Interop Deutsche Bank AG
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the 'License');
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an 'AS IS' BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { AsyncHandler } from '../AsyncHandler';
import { TransportConnection } from '@plexus-interop/transport-common';
import { Logger, LoggerFactory } from '@plexus-interop/common';
import { Completion, ErrorCompletion, ClientError } from '@plexus-interop/protocol';
import { AppLifeCycleManager } from '../lifecycle/AppLifeCycleManager';
import { ClientRequestProcessor } from './ClientRequestProcessor';
import { ApplicationConnection } from '../lifecycle/ApplicationConnection';
import { AuthenticationHandler } from './AuthenticationHandler';
import { TasksTracker } from './TasksTracker';

export class ClientConnectionProcessor implements AsyncHandler<TransportConnection, Completion> {

    constructor(
        private readonly authenticationProcessor: AuthenticationHandler,
        private readonly clientRequestProcessor: ClientRequestProcessor,
        private readonly appLifeCycleManager: AppLifeCycleManager) { }

    public async handle(connection: TransportConnection): Promise<Completion> {

        const connectionGuid = connection.uuid().toString();
        const log: Logger = LoggerFactory.getLogger(`ConnectionProcessor [${connectionGuid}]`);
        log.debug(`Received connection`);

        const requestsTracker = new TasksTracker();

        return new Promise((resolve, reject) => {
            let sourceConnection: undefined | ApplicationConnection;
            connection.subscribeToChannels({
                next: async channel => {
                    const channelStrId = channel.uuid().toString();
                    log.debug(`Received new channel [${channelStrId}]`);
                    if (!sourceConnection) {
                        try {
                            log.debug('First channel, trying to setup connection');
                            const appDescriptor = await this.authenticationProcessor.handle([connection, channel]);
                            const appConnection = await this.appLifeCycleManager.acceptConnection(connection, {
                                applicationId: appDescriptor.applicationId as string,
                                instanceId: appDescriptor.instanceId
                            }, c => {
                                log.error('Connection dropped');
                            });
                            sourceConnection = appConnection;
                            log.trace('Connected to client');
                        } catch (error) {
                            log.error('Unable to authenticate client connection', error);
                            connection.disconnect(new ErrorCompletion(new ClientError(error)));
                            reject(error);
                        }
                    } else {
                        log.trace(`Processing client request channel [${channelStrId}]`);
                        try {
                            const channelCompletion = await requestsTracker.start(
                                channelStrId,
                                () => this.clientRequestProcessor.handle(channel, sourceConnection as ApplicationConnection));
                            if (log.isTraceEnabled()) {
                                log.trace(`Received channel completion [${JSON.stringify(channelCompletion)}]`);
                            }
                        } catch (error) {
                            log.trace(`Failed on processing of [${channelStrId}] client request`);
                        }
                    }
                },
                complete: async () => {
                    log.debug(`Source connection completed`);
                    this.completeAndDisconnect(connection, requestsTracker, log)
                        .then(result => resolve(result))
                        .catch(e => {
                            log.error(`Failed to complete pending requests`, e);
                            reject(new ErrorCompletion(e));
                        });
                },
                error: e => {
                    log.error(`Error received from source connection`, e);
                    this.completeAndDisconnect(connection, requestsTracker, log)
                        .catch(completeErr => log.error('Failed to complete pending requests', e))
                        .then(() => log.debug('Pending requests completed'));
                    reject(new ErrorCompletion(e));
                }
            });
        });
    }

    private async completeAndDisconnect(connection: TransportConnection, requestsTracker: TasksTracker, log: Logger): Promise<Completion> {
        const result = await requestsTracker.completePending();
        log.info(`Completed pending actions`);
        await connection.disconnect(result);
        log.info(`Diconnected`);
        return result;
    }

}